sound/soc/sof/intel/hda-ipc.c

```diff
@@ -116,6 +116,12 @@ void hda_dsp_ipc_get_reply(struct snd_sof_dev *sdev)
 	spin_unlock_irqrestore(&sdev->ipc_lock, flags);
 }
 
+static bool hda_dsp_ipc_is_sof(uint32_t msg)
+{
+	return (msg & (HDA_DSP_IPC_PURGE_FW | 0xf << 9)) != msg ||
+		(msg & HDA_DSP_IPC_PURGE_FW) != HDA_DSP_IPC_PURGE_FW;
+}
```

Comment:
the test looks odd, not sure I get why the same PURGE_FW is tested twice?

Reply:
I think it is correct. You have to basically check for two things: that all HDA_DSP_IPC_PURGE_FW bits are set and no other bits (apart from 0xf << 9) are set. That's why it has two tests. The test *for* a ROM message is
```
(msg & (HDA_DSP_IPC_PURGE_FW | 0xf << 9)) == msg &&
(msg & HDA_DSP_IPC_PURGE_FW) == HDA_DSP_IPC_PURGE_FW
```
HDA_DSP_IPC_PURGE_FW | 0xf << 9 is the mask for all ROM messages. So, the first test checks, that msg has no bits set outside of that mask. The second test checks, that all PURGE bits are set in msg. The test in the function is just an inversion of this test. I haven't found an easier or more elegant form to perform this test. I'd certainly love to see one.